labAdmin/models.py

```diff
@@ -1,175 +1,157 @@
-
-        return ss if len(ss) > 0 else ""
-
-    def weekdays_permission_binary(self):
-        p = 0
-        p += 1<<0 if self.monday else 0
-        p += 1<<1 if self.tuesday else 0
-        p += 1<<2 if self.wednesday else 0
-        p += 1<<3 if self.thursday else 0
-        p += 1<<4 if self.friday else 0
-        p += 1<<5 if self.saturday else 0
-        p += 1<<6 if self.sunday else 0
-
-        return p
-
-    def weekdays_permission_binary_str(self):
-        return "{0:07b}".format(self.weekdays_permission_binary())
+    firstSignup = models.DateField()
+    lastSignup = models.DateField()
+    endSubcription = models.DateField()
+    needSubcription = models.BooleanField(default=True)
+    nfcId=models.BigIntegerField(unique=True, null=True)
+
+    # define Many-To-Many fields
+    groups=models.ManyToManyField('Group')
+
+    def subscription_end(self):
+        return self.endSubcription if self.needSubcription else "Subcription not needed"
+
+    def subscriptionExpired(self):
+        return self.needSubcription and self.endSubcription < timezone.now()
+
+    def can_open_door_now(self):
+        # Define groups and role
+        n = timezone.now()
+        return len(TimeSlot.objects.filter(role__group__user=self,role__role_kind=0,role__valid=True,hour_start__lte=n.time(),hour_end__gte=n.time(),weekday_start__lte=n.weekday(),weekday_end__gte=n.weekday())) > 0
+
+    def can_use_device_now(self, device):
+        n = timezone.now()
+        return len(TimeSlot.objects.filter(role__group__user=self,role__role_kind=1, role__category_device=device.category_device, role__valid=True,hour_start__lte=n.time(),hour_end__gte=n.time(),weekday_start__lte=n.weekday(),weekday_end__gte=n.weekday())) > 0
 
     def __str__(self):
         return self.name
 
-class Device(models.Model):
-    name=models.CharField(max_length=200)
-    dtype=models.ForeignKey('Devicetype')
-    hourlyCost=models.FloatField(default=0.0)
+class Category_Device(models.Model):
+    name=models.CharField(max_length=50)
 
     def __str__(self):
-        return self.name
+        return "%s"%(self.name)
 
-class Devicetype(models.Model):
-    name=models.CharField(max_length=200)
+class Group(models.Model):
+    name=models.CharField(max_length=50)
+    # define Many-To-Many fields
+    roles=models.ManyToManyField('Role')
+
+    def can_open_door_now(self):
+        # Define groups and role
+        n = timezone.now()
+        return len(TimeSlot.objects.filter(role__group=self,role__role_kind=0,role__valid=True,hour_start__lte=n.time(),hour_end__gte=n.time(),weekday_start__lte=n.weekday(),weekday_end__gte=n.weekday())) > 0
+
+    def can_use_device_now(self, device):
+        n = timezone.now()
+        return len(TimeSlot.objects.filter(role__group=self,role__role_kind=1, role__category_device=device.category_device, role__valid=True,hour_start__lte=n.time(),hour_end__gte=n.time(),weekday_start__lte=n.weekday(),weekday_end__gte=n.weekday())) > 0
```

Comment:
This is very risky because an exception can open the door:

```
>>> ValueError('stuff2 not in content2') > 0
True
```
We need a try except here. same at line 107